### PR TITLE
event-auditor: ebpf-manager: centralize ebpf management logic

### LINKED ISSUE #334: event-auditor: current ebpf management is only focused on maps
## Redesign needed (bug/enhancement)

**Short Description**

The current (event-auditor branch) [sharedMapManager.go](https://github.com/kubearmor/KubeArmor/blob/event-auditor/KubeArmor/eventAuditor/sharedMapManager.go) contains ebpf related code mainly focused on managing localized maps. eventAuditor requires more than that, like ebpf object, program and link management.

**Is your feature request related to a problem? Please describe the use case.**

https://github.com/kubearmor/KubeArmor/blob/5b659e4f96a4e8288c86493f235ade9a64d8f92e/KubeArmor/eventAuditor/sharedMapManager.go#L21-L24

In the absence of object management in the `SharedMapManager` (and program, link), an `InitMap`, for example, may incorrectly load maps from a same object into memory more than once. 

https://github.com/kubearmor/KubeArmor/blob/5b659e4f96a4e8288c86493f235ade9a64d8f92e/KubeArmor/eventAuditor/sharedMapManager.go#L67

The same lack can erroneously destroy all maps, programs and links related to an object to be closed.

https://github.com/kubearmor/KubeArmor/blob/5b659e4f96a4e8288c86493f235ade9a64d8f92e/KubeArmor/eventAuditor/sharedMapManager.go#L118

**The solution is a WIP**

`KABPFManager` is being implemented as a generic replacement that can be used throughout the project.

```go
type KABPFManager struct {
	objs          map[KABPFObjFileName]*lbpf.KABPFObject
	maps          map[KABPFMapName]*lbpf.KABPFMap
	progs         map[KABPFProgName]*lbpf.KABPFProgram
	links         map[KABPFLinkName]*lbpf.KABPFLink
	objsMapsPath  string
	objsProgsPath string
}
```

### PR DESCRIPTION
**Pull request by geyslan**
This PR does a lot of changes to the current ebpf management.

The new KABPFManager can manage programs and links as well.

Summary:

- [x] Change EventAuditor and processSpecHandler.go interface functions  to use the new KABPFManager
- [x] Update sharedMaps.go to be properly consumed by KABPFManager
- [x] Introduce processProgs.go to be consumed by KABPFManager
- [x] Add initial ebpf program "ka_ea_process_prog.bpf.c".

- [x] fix typos; fix license comment order; update go.mod go.sum; remove deprecated files

Fixes #334

### CODE REVIEW

File: KubeArmor/eventAuditor/ebpfManager.go [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+		return err
+	}
+
+	delete(bm.links, kaProg.EventName)
+
+	return err
+}
+
+// isObjCloseable Function
+func (bm *KABPFManager) isObjCloseable(o *lbpf.KABPFObject) bool {
```
Comment (geyslan) [inline KubeArmor/eventAuditor/ebpfManager.go:225]: This function is a rough implementation. We can change it later to check an object reference counter.

File: KubeArmor/eventAuditor/ebpfManager.go [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+	if p = bm.getProg(kaProg.Name); p == nil {
+		return fmt.Errorf("program %v not initialized", kaProg.Name)
+	}
+
+	if l = bm.getLink(kaProg.EventName); l != nil {
+		return fmt.Errorf("program %v already attached to %v", kaProg.Name, kaProg.EventName)
+	}
+
+	switch kaProg.EventType {
+	case lbpf.KABPFLinkTypeTracepoint:
```
Comment (geyslan) [inline KubeArmor/eventAuditor/ebpfManager.go:187]: If this approach is accepted we can add the other attach types.

File: KubeArmor/BPF/ka_ea_process_prog.bpf.c [hunk 3]
```diff
@@ -0,0 +1,25 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright 2021 Authors of KubeArmor
+
+#include "vmlinux.h"
+#include <bpf/bpf_helpers.h>
+
+#include "maps.bpf.h"
+
+SEC("tracepoint/sys_enter_execve")
```
Comment (nam-jaehyun) [inline KubeArmor/BPF/ka_ea_process_prog.bpf.c:9]: in general, we used kprobe. is there any difference between tracepoint and kprobe?

Comment (geyslan) [reply KubeArmor/BPF/ka_ea_process_prog.bpf.c:9]: Tracepoints are static therefore more stable:

https://stackoverflow.com/a/47031794

https://blogs.oracle.com/linux/post/taming-tracepoints-in-the-linux-kernel

https://www.oreilly.com/library/view/linux-observability-with/9781492050193/ch04.html

**Pull request merged**